### PR TITLE
Feature: pause hydration when image verification fails

### PR DESCRIPTION
**Pull request by tiffanny29631**

### CODE REVIEW

File: manifests/templates/reconciler-manager-configmap.yaml [hunk 1]
```diff
@@ -70,10 +70,14 @@ data:
            - name: service-account # This is needed to prevent any credentials from being exposed to malicious users.
              mountPath: /var/run/secrets/kubernetes.io/serviceaccount
              readOnly: true
+           - name: shared-volume  # Mount the shared volume for reading
+             mountPath: /shared
```
Comment (nan-yu) [inline manifests/templates/reconciler-manager-configmap.yaml:74]: The volume name is too general. It seems the volume is supposed to be written by the reconciler and read by the hydration-controller. I would suggest renaming to something like `/reconciler-signals` to make it self-explanatory.

File: manifests/templates/reconciler-manager-configmap.yaml [hunk 2]
```diff
@@ -265,6 +271,8 @@ data:
            emptyDir: {}
          - name: kube
            emptyDir: {}
+         - name: shared-volume
+           emptyDir: {}  # Shared volume for ready-to-render file
```
Comment (nan-yu) [inline manifests/templates/reconciler-manager-configmap.yaml:275]: `Shared volume for ready-to-render file` -> `A shared volume that allows the reconciler to send signals to the hydration-controller`.

File: pkg/hydrate/controller.go [hunk 3]
```diff
@@ -100,12 +103,25 @@ func (h *Hydrator) Run(ctx context.Context) {
 						srcCommit, err)
 				}
 			} else if doneCommit(h.DonePath.OSPath()) != srcCommit {
-				// If the commit has been processed before, regardless of success or failure,
-				// skip the hydration to avoid repeated execution.
-				// The rehydrate ticker will retry on the failed commit.
-				hydrateErr = h.hydrate(srcCommit, syncPath)
-				if err := h.complete(srcCommit, hydrateErr); err != nil {
-					klog.Errorf("failed to complete the rendering execution for commit %q: %v", srcCommit, err)
+				absShared := cmpath.Absolute("/shared")
```
Comment (nan-yu) [inline pkg/hydrate/controller.go:106]: Either define the volume path as a constant, or define it as a flag (similar to other root directories).

File: pkg/hydrate/controller.go [hunk 4]
```diff
@@ -100,12 +103,25 @@ func (h *Hydrator) Run(ctx context.Context) {
...
+				readyToRenderFile := absShared.Join(cmpath.RelativeSlash(ReadyToRenderFile)).OSPath()
+				readyToRenderCommit, err := ExtractCommit(readyToRenderFile)
+				if err != nil {
+					klog.Error(err)
+					h.complete(readyToRenderCommit, NewActionableError(err))
```
Comment (nan-yu) [inline pkg/hydrate/controller.go:111]: If the ready-to-render file doesn't exist, should we mark the hydration complete?
`h.complete` updates the done file and the rendering error file. 
In this case, no rendering is being executed. Moreover, if the commit is not ready to render, the reconciler won't move forward with checking the rendering state.

I think we can simply log a warning in the hydration-controller.
Then this can be simplified as
```
else if readyToRenderFile(h.ReadyToRenderPath.OSPath()) == srcCommit && doneCommit(h.DonePath.OSPath()) != srcCommit {
   .... 
}

Basically, we only run rendering when the commit is ready and the commit is not rendered before.
```

**Review by nan-yu — COMMENTED** [review]
Is it possible to add unit tests for the blocking feature?

File: pkg/parse/run.go [hunk 5]
```diff
@@ -120,6 +120,13 @@ func DefaultRunFunc(ctx context.Context, r Reconciler, trigger string) RunResult
 	if gs.Errs == nil {
 		if err := r.SyncStatusClient().SetImageToSyncAnnotation(ctx, gs.Commit); err != nil {
```
Comment (nan-yu) [inline pkg/parse/run.go:121]: A comment that is not specific to the unblocking feature. Now that we have `ReconcilerState.ReconcilerStatus` populated from RSync, we can skip setting the annotation and updating the ready-to-render file when the commit is not changed.

In `type ReconcilerStatus struct {}`, we can add a field `ImageToSyncAnnotation` to cache the annotation, and here we update `if gs.Errs == nil {` to `if gs.Errs == nil && r.ReconcilerState().status.ImageToSyncAnnotation != gs.Commit {`.

This can be a separate PR.

File: manifests/templates/reconciler-manager-configmap.yaml (hunk 1 above)
Comment (tiffanny29631) [reply manifests/templates/reconciler-manager-configmap.yaml]: Done

File: manifests/templates/reconciler-manager-configmap.yaml (hunk 2 above)
Comment (tiffanny29631) [reply manifests/templates/reconciler-manager-configmap.yaml]: Done

File: pkg/hydrate/controller.go (hunk 4 above)
Comment (tiffanny29631) [reply pkg/hydrate/controller.go]: Done, only log a warning if hydration is skipped due to unexpected readyToRenderCommit.

**Comment by tiffanny29631** [comment]
>Is it possible to add unit tests for the blocking feature?

Working on an e2e test now. The check of readyToRenderCommit happens in the controller Run function so is not easy to emulate, we could check for the warning message in hydration-controller log when commit is not as expected, and pass when an image with dry config syncs correctly - meaning hydration is resumed.

**Comment by tiffanny29631** [comment]
@nan-yu e2e test in place, ready for another look.

File: pkg/hydrate/controller.go [hunk 6]
```diff
@@ -99,20 +104,34 @@ func (h *Hydrator) Run(ctx context.Context) {
...
+					// skip the hydration to avoid repeated execution.
+					// The rehydrate ticker will retry on the failed commit.
+					// Only proceed to rendering if the ready-to-render file contains the srcCommit
+					hydrateErr = h.hydrate(srcCommit, syncPath)
+					if err := h.complete(srcCommit, hydrateErr); err != nil {
+						klog.Errorf("failed to complete the rendering execution for commit %q: %v", srcCommit, err)
+					}
+				} else {
+					klog.Warningf("skip hydration as readyToRenderCommit does not match srcCommit, want %s, got %s", srcCommit, readyToRenderCommit)
 				}
```
Comment (nan-yu) [inline pkg/hydrate/controller.go:120]: Suggest changing to a switch block to make it clear.

```
else {
  doneCommit := extractCommit(h.DonePath.OSPath())
  readyToRenderFile := h.ReconcilerSignalDir.Join(cmpath.RelativeSlash(ReadyToRenderFile)).OSPath()
  readyToRenderCommit := extractCommit(readyToRenderFile)
  switch {
  case doneCommit == srcCommit:
    // no-op
    // If the commit has been processed before, skip the hydration to avoid repeated execution regardless of success or failure.
	// The rehydrate ticker will retry on the failed commit.
  case doneCommit != srcCommit && readyToRenderCommit != srcCommit:
    klog.Warningf("skip hydration as readyToRenderCommit does not match srcCommit, want %s, got %s", srcCommit, readyToRenderCommit)
  case doneCommit != srcCommit && readyToRenderCommit == srcCommit:
    // Only proceed to rendering if the ready-to-render file contains the srcCommit
	hydrateErr = h.hydrate(srcCommit, syncPath)
	if err := h.complete(srcCommit, hydrateErr); err != nil {
	  klog.Errorf("failed to complete the rendering execution for commit %q: %v", srcCommit, err)
	}
  }
}
```

File: pkg/parse/run.go [hunk 7]
```diff
@@ -132,6 +132,18 @@ func (r *reconciler) Reconcile(ctx context.Context, trigger string) ReconcileRes
 		state.RecordFailure(opts.Clock, errs)
 		return result
 	}
+	if newSourceStatus.Errs == nil {
```
[the remark this answers is not in the document]
Comment (nan-yu) [reply pkg/parse/run.go:135]: The code only writes the ready-to-render file. Do we need to remove it when there is source error?

File: pkg/hydrate/controller.go (hunk 6 above)
Comment (tiffanny29631) [reply pkg/hydrate/controller.go:124]: Done

File: pkg/parse/run_test.go [hunk 8]
```diff
@@ -1135,6 +1146,12 @@ func TestReconciler_Reconcile(t *testing.T) {
 			require.NoError(t, err)
 			testutil.AssertEqual(t, tc.expectedRootSyncFunc(sourceRoot), rs)
 
+			if tc.imageVerified {
+				readyToRenderFilePath := filepath.Join(reconcilerSignalDir, "ready-to-render")
+				_, err = os.Stat(readyToRenderFilePath)
```
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply pkg/parse/run_test.go:1151]: Done

File: pkg/parse/run.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply pkg/parse/run.go]: Moved the update of ready-to-render in the fetch action, as it's tightly coupled with the result of setting image-to-sync annotation.

File: manifests/templates/reconciler-manager-configmap.yaml [hunk 9]
```diff
@@ -64,16 +64,21 @@ data:
...
              mountPath: /var/run/secrets/kubernetes.io/serviceaccount
              readOnly: true
+           - name: reconciler-signals
+             mountPath: /reconciler-signals
+             readOnly: true
            imagePullPolicy: IfNotPresent
            securityContext:
              allowPrivilegeEscalation: false
              readOnlyRootFilesystem: false
+
```
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply manifests/templates/reconciler-manager-configmap.yaml]: Done.

File: e2e/testcases/otel_collector_test.go
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply e2e/testcases/otel_collector_test.go:1]: Done

File: e2e/testcases/oci_signature_verification_test.go
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply e2e/testcases/oci_signature_verification_test.go:1]: Changed to use wet repo as hydration is not verified in e2e.

File: e2e/testdata/hydration/kustomize-components-more/base/kustomization.yaml
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply e2e/testdata/hydration/kustomize-components-more/base/kustomization.yaml:1]: Changed to use wet repo as hydration is not verified in e2e.

File: e2e/testdata/hydration/namespace-repo-more/base/kustomization.yaml
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply e2e/testdata/hydration/namespace-repo-more/base/kustomization.yaml:1]: Changed to use wet repo as hydration is not verified in e2e.

File: cmd/reconciler/main.go [hunk 10]
```diff
@@ -102,27 +102,31 @@ var (
 
 	dynamicNSSelectorEnabled = flag.Bool("dynamic-ns-selector-enabled", util.EnvBool(reconcilermanager.DynamicNSSelectorEnabled, false), "")
 
-	webhookEnabled = flag.Bool("webhook-enabled", util.EnvBool(reconcilermanager.WebhookEnabled, false), "")
+	webhookEnabled       = flag.Bool("webhook-enabled", util.EnvBool(reconcilermanager.WebhookEnabled, false), "")
+	reconcilerSignalsDir = flag.String(flags.reconcilerSignalDir, "/reconciler-signals",
+		"The absolute path in the container that contains the image digest which has passed signature verification")
```
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply cmd/reconciler/main.go]: Done

File: pkg/hydrate/controller.go [hunk 11]
```diff
@@ -99,20 +104,38 @@ func (h *Hydrator) Run(ctx context.Context) {
...
 			}
 			runTimer.Reset(h.PollingPeriod) // Schedule re-run attempt
 		}
 	}
 }
 
+// isReadyToRender returns true if commit in ready-to-render file is the same as
+// commit acquired from source
+func (h *Hydrator) isReadyToRender(srcCommit string) (string, bool) {
```
Comment (nan-yu) [inline pkg/hydrate/controller.go:133]: Is this function still needed?

File: pkg/parse/run.go [hunk 12]
```diff
@@ -218,6 +218,13 @@ func (r *reconciler) fetch(ctx context.Context) (*SourceStatus, cmpath.Absolute,
 	if newSourceStatus.Errs == nil {
 		if err := r.syncStatusClient.SetImageToSyncAnnotation(ctx, newSourceStatus.Commit); err != nil {
 			newSourceStatus.Errs = status.Append(newSourceStatus.Errs, err)
+		} else {
+			// write the commit into the read-to-render file in hydrated root
```
Comment (nan-yu) [inline pkg/parse/run.go:222]: typo: `ready-to-render`

File: pkg/parse/run.go [hunk 13]
```diff
@@ -218,6 +218,13 @@ func (r *reconciler) fetch(ctx context.Context) (*SourceStatus, cmpath.Absolute,
...
+			if opts.RenderingEnabled {
```
Comment (nan-yu) [inline pkg/parse/run.go:223]: nit: this `if` block can be combined with the `else` block -> `else if opts.RenderingEnabled {`

File: pkg/parse/run.go [hunk 14]
```diff
@@ -218,6 +218,13 @@ func (r *reconciler) fetch(ctx context.Context) (*SourceStatus, cmpath.Absolute,
 	if newSourceStatus.Errs == nil {
```
Comment (nan-yu) [inline pkg/parse/run.go:218]: Do we need to remove the `ready-to-render` file when there is source error?

File: pkg/hydrate/controller.go (hunk 11 above)
Comment (tiffanny29631) [reply pkg/hydrate/controller.go:133]: Yes, I split it out so that we can do unit test on it.

File: pkg/parse/run.go (hunk 12 above)
Comment (tiffanny29631) [reply pkg/parse/run.go]: Done

File: pkg/parse/run.go (hunk 14 above)
Comment (tiffanny29631) [reply pkg/parse/run.go:218]: Added logic to remove this file when a source error occurs, either before or after image verification. If the removal fails, it logs the error without appending it to the source error or interrupting any loop. Since the hydration controller won’t receive the expected commit in the ready-to-render file when a source error exists, rendering is effectively blocked regardless.

File: pkg/hydrate/controller.go (hunk 11 above)
Comment (nan-yu) [reply pkg/hydrate/controller.go:133]: It is only used in the unit test. If it is not used by the controller, we don't need to unit test this function.

File: pkg/parse/run.go [hunk 15]
```diff
@@ -615,6 +616,24 @@ func unblockHydration(commit string, r Reconciler) status.Error {
...
+// file will not be updated with the expected commit that is expected by
+// hydration-controller, thus rendering will be blocked.
+func blockHydration(r Reconciler) {
+	readyToRenderFile := r.Options().ReconcilerSignalsDir.Join(cmpath.RelativeSlash(hydrate.ReadyToRenderFile)).OSPath()
+	err := os.Remove(readyToRenderFile)
+	if err != nil {
+		if os.IsNotExist(err) {
+			klog.Infof("ready-to-render file %s not found, no action needed", readyToRenderFile)
+		}
+		klog.Errorf("failed to removed ready-to-render file %s: %v", readyToRenderFile, err)
```
Comment (nan-yu) [inline pkg/parse/run.go:632]: typo: removed -> remove

File: pkg/parse/run.go [hunk 16]
```diff
@@ -615,6 +616,24 @@ func unblockHydration(commit string, r Reconciler) status.Error {
...
+// "file not found". This function removes the ready-to-render file at its
+// best effort, since with source errors presenting, the ready-to-render
+// file will not be updated with the expected commit that is expected by
+// hydration-controller, thus rendering will be blocked.
+func blockHydration(r Reconciler) {
+	readyToRenderFile := r.Options().ReconcilerSignalsDir.Join(cmpath.RelativeSlash(hydrate.ReadyToRenderFile)).OSPath()
+	err := os.Remove(readyToRenderFile)
+	if err != nil {
+		if os.IsNotExist(err) {
+			klog.Infof("ready-to-render file %s not found, no action needed", readyToRenderFile)
```
Comment (nan-yu) [inline pkg/parse/run.go:630]: nit: this log can be set at V(3) level.

File: pkg/parse/run.go [hunk 17]
```diff
@@ -584,6 +592,48 @@ func (r *reconciler) startAsyncStatusUpdates(ctx context.Context) <-chan struct{
...
+}
+
+// blockHydration removes the ready-to-render file from the shared
+// directory and returns any error that occurs during the procedure except
+// "file not found". This function removes the ready-to-render file at its
+// best effort, since with source errors presenting, the ready-to-render
+// file will not be updated with the expected commit that is expected by
+// hydration-controller, thus rendering will be blocked.
+func blockHydration(r Reconciler) {
+	readyToRenderFile := r.Options().ReconcilerSignalsDir.Join(cmpath.RelativeSlash(hydrate.ReadyToRenderFile)).OSPath()
```
Comment (nan-yu) [inline pkg/parse/run.go:626]: This is repeated in both unblock and block functions. Can it be extracted to a variable?

File: pkg/parse/run.go [hunk 18]
```diff
@@ -584,6 +592,48 @@ func (r *reconciler) startAsyncStatusUpdates(ctx context.Context) <-chan struct{
 	return doneCh
 }
 
+// unblockHydration adds the image digest into the ready-to-render file under the
+// shared directory between reconciler and hydration-controller to inform the
+// hydration-controller to proceed rendering
+func unblockHydration(commit string, r Reconciler) status.Error {
+	readyToRenderFile := r.Options().ReconcilerSignalsDir.Join(cmpath.RelativeSlash(hydrate.ReadyToRenderFile)).OSPath()
+	klog.Infof("writing commit %s to ready to render file %s", commit, readyToRenderFile)
```
Comment (nan-yu) [inline pkg/parse/run.go:600]: nit: `writing commit %s to ready to render file` -> `signaling the hydration-controller that commit %s is ready to render`

File: pkg/hydrate/controller.go (hunk 11 above)
Comment (tiffanny29631) [reply pkg/hydrate/controller.go]: Ah I forgot the logic in Run has changed. Just updated with cleanup.

**Review by nan-yu — APPROVED** [review]
/lgtm
